// mat-video/esm5/app/video/ui/mat-frame-by-frame-control/mat-frame-by-frame-control.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
var MatFrameByFrameControlComponent = /** @class */ (function () {
    function MatFrameByFrameControlComponent() {
        this.fps = 29.97;
    }
    MatFrameByFrameControlComponent.prototype.ngOnInit = function () {
    };
    MatFrameByFrameControlComponent.prototype.seekFrames = function (nbFrames) {
        if (!this.video.paused) {
            this.video.pause();
        }
        var currentFrames = this.video.currentTime * this.fps;
        var newPos = ((currentFrames + nbFrames) / this.fps) + 0.00001;
        this.video.currentTime = newPos;
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", HTMLVideoElement)
    ], MatFrameByFrameControlComponent.prototype, "video", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], MatFrameByFrameControlComponent.prototype, "fps", void 0);
    MatFrameByFrameControlComponent = tslib_1.__decorate([
        Component({
            selector: 'mat-frame-by-frame-control',
            template: "<button mat-icon-button (click)=\"seekFrames(-5)\">\r\n    <mat-icon>skip_previous</mat-icon>\r\n</button>\r\n\r\n<button mat-icon-button (click)=\"seekFrames(-1)\">\r\n    <mat-icon>arrow_left</mat-icon>\r\n</button>\r\n\r\n<button mat-icon-button (click)=\"seekFrames(1)\">\r\n    <mat-icon>arrow_right</mat-icon>\r\n</button>\r\n\r\n<button mat-icon-button (click)=\"seekFrames(5)\">\r\n    <mat-icon>skip_next</mat-icon>\r\n</button>",
            styles: [""]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], MatFrameByFrameControlComponent);
    return MatFrameByFrameControlComponent;
}());
export { MatFrameByFrameControlComponent };
