// mat-video/esm5/app/video/directives/mat-video-track.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Host, Input } from '@angular/core';
import { MatVideoComponent } from '../video.component';
var MatVideoTrackDirective = /** @class */ (function () {
    function MatVideoTrackDirective(matVideoComponent, el) {
        this.matVideoComponent = matVideoComponent;
        this.el = el;
        this.src = null;
        this.kind = null;
        this.srclang = null;
        this.label = null;
        this.init = true;
        this.video = matVideoComponent;
        this.track = el.nativeElement;
        this.init = false;
    }
    MatVideoTrackDirective.prototype.ngOnChanges = function (changes) {
        this.track.src = this.src;
        this.track.kind = this.kind;
        this.track.srclang = this.srclang;
        this.track.label = this.label;
        if (!this.init)
            this.video.load();
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], MatVideoTrackDirective.prototype, "src", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], MatVideoTrackDirective.prototype, "kind", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], MatVideoTrackDirective.prototype, "srclang", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], MatVideoTrackDirective.prototype, "label", void 0);
    MatVideoTrackDirective = tslib_1.__decorate([
        Directive({
            selector: '[matVideoTrack]'
        }),
        tslib_1.__param(0, Host()),
        tslib_1.__metadata("design:paramtypes", [MatVideoComponent,
            ElementRef])
    ], MatVideoTrackDirective);
    return MatVideoTrackDirective;
}());
export { MatVideoTrackDirective };
