// ngx-editor/esm5/lib/ngx-editor.service.js
import { __decorate, __param } from "tslib";
import { Injectable, Optional } from '@angular/core';
import menu from './prosemirror/plugins/menu';
import placeholder from './prosemirror/plugins/placeholder';
import { schema } from './schema';
import * as i0 from "@angular/core";
var NgxEditorServiceConfig = /** @class */ (function () {
    function NgxEditorServiceConfig() {
        this.plugins = [
            menu(),
            placeholder()
        ];
        this.nodeViews = {};
        this.schema = schema;
    }
    NgxEditorServiceConfig.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxEditorServiceConfig_Factory() { return new NgxEditorServiceConfig(); }, token: NgxEditorServiceConfig, providedIn: "root" });
    NgxEditorServiceConfig = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], NgxEditorServiceConfig);
    return NgxEditorServiceConfig;
}());
export { NgxEditorServiceConfig };
var NgxEditorService = /** @class */ (function () {
    function NgxEditorService(config) {
        this.config = config;
    }
    NgxEditorService.ctorParameters = function () { return [
        { type: NgxEditorServiceConfig, decorators: [{ type: Optional }] }
    ]; };
    NgxEditorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxEditorService_Factory() { return new NgxEditorService(i0.ɵɵinject(NgxEditorServiceConfig, 8)); }, token: NgxEditorService, providedIn: "root" });
    NgxEditorService = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __param(0, Optional())
    ], NgxEditorService);
    return NgxEditorService;
}());
export { NgxEditorService };
var defaultConfig = {
    plugins: [],
    nodeViews: {},
    schema: schema
};
export function provideMyServiceOptions(config) {
    return Object.assign({}, defaultConfig, config);
}
