// ngx-editor/esm5/lib/ngx-editor.module.js
import { __decorate } from "tslib";
import { NgModule, InjectionToken } from '@angular/core';
import { NgxEditorComponent } from './ngx-editor.component';
import { NgxEditorServiceConfig, provideMyServiceOptions } from './ngx-editor.service';
var NGX_EDITOR_CONFIG_TOKEN = new InjectionToken('NgxEditorConfig');
var NgxEditorModule = /** @class */ (function () {
    function NgxEditorModule() {
    }
    NgxEditorModule_1 = NgxEditorModule;
    NgxEditorModule.forRoot = function (config) {
        return {
            ngModule: NgxEditorModule_1,
            providers: [
                {
                    provide: NGX_EDITOR_CONFIG_TOKEN,
                    useValue: config
                },
                {
                    provide: NgxEditorServiceConfig,
                    useFactory: provideMyServiceOptions,
                    deps: [NGX_EDITOR_CONFIG_TOKEN]
                }
            ]
        };
    };
    var NgxEditorModule_1;
    NgxEditorModule = NgxEditorModule_1 = __decorate([
        NgModule({
            declarations: [NgxEditorComponent],
            exports: [NgxEditorComponent],
        })
    ], NgxEditorModule);
    return NgxEditorModule;
}());
export { NgxEditorModule };
