// ngx-editor/esm2015/lib/ngx-editor.service.js
import { __decorate, __param } from "tslib";
import { Injectable, Optional } from '@angular/core';
import menu from './prosemirror/plugins/menu';
import placeholder from './prosemirror/plugins/placeholder';
import { schema } from './schema';
import * as i0 from "@angular/core";
let NgxEditorServiceConfig = class NgxEditorServiceConfig {
    constructor() {
        this.plugins = [
            menu(),
            placeholder()
        ];
        this.nodeViews = {};
        this.schema = schema;
    }
};
NgxEditorServiceConfig.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxEditorServiceConfig_Factory() { return new NgxEditorServiceConfig(); }, token: NgxEditorServiceConfig, providedIn: "root" });
NgxEditorServiceConfig = __decorate([
    Injectable({
        providedIn: 'root'
    })
], NgxEditorServiceConfig);
export { NgxEditorServiceConfig };
let NgxEditorService = class NgxEditorService {
    constructor(config) {
        this.config = config;
    }
};
NgxEditorService.ctorParameters = () => [
    { type: NgxEditorServiceConfig, decorators: [{ type: Optional }] }
];
NgxEditorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxEditorService_Factory() { return new NgxEditorService(i0.ɵɵinject(NgxEditorServiceConfig, 8)); }, token: NgxEditorService, providedIn: "root" });
NgxEditorService = __decorate([
    Injectable({
        providedIn: 'root'
    }),
    __param(0, Optional())
], NgxEditorService);
export { NgxEditorService };
const defaultConfig = {
    plugins: [],
    nodeViews: {},
    schema
};
export function provideMyServiceOptions(config) {
    return Object.assign({}, defaultConfig, config);
}
