// ngx-editor/esm2015/lib/ngx-editor.module.js
var NgxEditorModule_1;
import { __decorate } from "tslib";
import { NgModule, InjectionToken } from '@angular/core';
import { NgxEditorComponent } from './ngx-editor.component';
import { NgxEditorServiceConfig, provideMyServiceOptions } from './ngx-editor.service';
const NGX_EDITOR_CONFIG_TOKEN = new InjectionToken('NgxEditorConfig');
let NgxEditorModule = NgxEditorModule_1 = class NgxEditorModule {
    static forRoot(config) {
        return {
            ngModule: NgxEditorModule_1,
            providers: [
                {
                    provide: NGX_EDITOR_CONFIG_TOKEN,
                    useValue: config
                },
                {
                    provide: NgxEditorServiceConfig,
                    useFactory: provideMyServiceOptions,
                    deps: [NGX_EDITOR_CONFIG_TOKEN]
                }
            ]
        };
    }
};
NgxEditorModule = NgxEditorModule_1 = __decorate([
    NgModule({
        declarations: [NgxEditorComponent],
        exports: [NgxEditorComponent],
    })
], NgxEditorModule);
export { NgxEditorModule };
