// ngx-markdown-editor/esm5/public_api.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { MarkdownEditorComponent } from './lib/md-editor.component';
import { MarkdownEditorResizeSensorComponent } from './lib/resize-sensor/resize-sensor.component';
export { MarkdownEditorComponent } from './lib/md-editor.component';
export { MarkdownEditorResizeSensorComponent } from './lib/resize-sensor/resize-sensor.component';
var LMarkdownEditorModule = /** @class */ (function () {
    function LMarkdownEditorModule() {
    }
    LMarkdownEditorModule = __decorate([
        NgModule({
            declarations: [
                MarkdownEditorComponent,
                MarkdownEditorResizeSensorComponent
            ],
            imports: [
                CommonModule,
                FormsModule
            ],
            exports: [
                MarkdownEditorComponent,
                MarkdownEditorResizeSensorComponent
            ]
        })
    ], LMarkdownEditorModule);
    return LMarkdownEditorModule;
}());
export { LMarkdownEditorModule };
