// ngx-markdown-editor/esm2015/lib/resize-sensor/resize-sensor.component.js
import { __decorate, __metadata } from "tslib";
import { Component, NgZone, Input, Output, ViewChild, ElementRef, EventEmitter } from '@angular/core';
let MarkdownEditorResizeSensorComponent = class MarkdownEditorResizeSensorComponent {
    constructor(_ngZone) {
        this._ngZone = _ngZone;
        this.interval = 500;
        this.resize = new EventEmitter();
        this.sizeInfo = {
            width: 0,
            height: 0
        };
    }
    ngAfterViewInit() {
        this.sizeInfo = {
            width: this.resizeSensor.nativeElement.offsetWidth,
            height: this.resizeSensor.nativeElement.offsetHeight
        };
        this._ngZone.runOutsideAngular(() => {
            setTimeout(() => {
                this.detectSize();
            }, this.interval);
        });
    }
    detectSize() {
        let width = this.resizeSensor.nativeElement.offsetWidth;
        let height = this.resizeSensor.nativeElement.offsetHeight;
        if (this.sizeInfo.width !== width || this.sizeInfo.height !== height) {
            this.sizeInfo = {
                width: width,
                height: height
            };
            this.resize.emit(this.sizeInfo);
        }
        setTimeout(() => {
            this.detectSize();
        }, this.interval);
    }
};
MarkdownEditorResizeSensorComponent.ctorParameters = () => [
    { type: NgZone }
];
__decorate([
    ViewChild('resizeSensor', { static: true }),
    __metadata("design:type", ElementRef)
], MarkdownEditorResizeSensorComponent.prototype, "resizeSensor", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], MarkdownEditorResizeSensorComponent.prototype, "interval", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], MarkdownEditorResizeSensorComponent.prototype, "resize", void 0);
MarkdownEditorResizeSensorComponent = __decorate([
    Component({
        selector: 'md-editor-resize-sensor',
        template: "<div class=\"md-editor-resize-sensor\" #resizeSensor></div>\r\n",
        styles: [".md-editor-resize-sensor{position:absolute;left:0;right:0;top:0;bottom:0;z-index:-1}"]
    }),
    __metadata("design:paramtypes", [NgZone])
], MarkdownEditorResizeSensorComponent);
export { MarkdownEditorResizeSensorComponent };
