// ngx-editor/esm5/lib/prosemirror/plugins/menu/menu.js
import { toggleMark } from 'prosemirror-commands';
import { isNodeActive, isMarkActive, isListItem } from '../../helpers';
import { toggleList, toggleBlockType, toggleWrap } from '../../commands';
import { getIconSvg } from '../../../utils/icons';
import flatDeep from '../../../utils/flatDeep';
import menuItemsMeta from './meta';
var SEPERATOR_CLASSNAME = 'NgxEditor__Seperator';
var MENU_ITEM_CLASSNAME = 'NgxEditor__MenuItem';
var ACTIVE_MENU_ITEM_CLASSNAME = MENU_ITEM_CLASSNAME + "--Active";
var DISABLED_CLASSNAME = 'NgxEditor--Disabled';
var DROPDWON_ITEM_CLASSNAME = 'NgxEditor__Dropdown';
var DROPWDOWN_OPEN_CLASSNAME = DROPDWON_ITEM_CLASSNAME + "--Open";
var ACTIVE_DROPDOWN_ITEM_CLASSNAME = DROPDWON_ITEM_CLASSNAME + "--Active";
var SELECTED_DROPDOWN_ITEM_CLASSNAME = DROPDWON_ITEM_CLASSNAME + "--Selected";
var DROPDOWN_ITEMS = new Map();
DROPDOWN_ITEMS.set('heading', ['h1', 'h2', 'h3', 'h4', 'h5', 'h6']);
var DropDownView = /** @class */ (function () {
    function DropDownView(dropdownGroup, dropdownFields, editorView, options) {
        this.updates = [];
        this.dropdownGroup = dropdownGroup;
        this.dropdownFields = dropdownFields;
        this.editorView = editorView;
        this.options = options;
    }
    DropDownView.prototype.getWrapperDom = function () {
        var _this = this;
        var isDropdownOpen = false;
        var dropdown = document.createElement('div');
        var labels = this.options.labels;
        dropdown.classList.add(DROPDWON_ITEM_CLASSNAME);
        var dropdownText = document.createElement('div');
        dropdownText.classList.add(DROPDWON_ITEM_CLASSNAME + "__Text");
        dropdownText.textContent = labels[this.dropdownGroup];
        dropdown.appendChild(dropdownText);
        // create dropdown list
        var dropdownMenu = document.createElement('div');
        dropdownMenu.classList.add(DROPDWON_ITEM_CLASSNAME + "__DropdownMenu");
        var mouseDownHandler = function (e) {
            e.preventDefault();
            if (!dropdown.contains(e.target)) {
                closeDropdown();
            }
        };
        var openDropdown = function (e) {
            var target = e.target;
            if (dropdownMenu.contains(target)) {
                return;
            }
            dropdown.classList.add(DROPWDOWN_OPEN_CLASSNAME);
            isDropdownOpen = true;
            window.addEventListener('mousedown', mouseDownHandler);
        };
        var closeDropdown = function () {
            dropdown.classList.remove(DROPWDOWN_OPEN_CLASSNAME);
            isDropdownOpen = false;
            window.removeEventListener('mousedown', mouseDownHandler);
        };
        dropdown.addEventListener('click', function (e) {
            e.preventDefault();
            if (!isDropdownOpen) {
                openDropdown(e);
            }
            else {
                closeDropdown();
            }
        });
        this.dropdownFields.forEach(function (dropdownItem) {
            var menuItem = menuItemsMeta[dropdownItem];
            var text = labels[menuItem.key];
            if (menuItem.key === 'heading') {
                text += " " + menuItem.attrs.level;
            }
            var spec = {
                classNames: [
                    DROPDWON_ITEM_CLASSNAME + "__Item"
                ],
                textContent: text,
                attrs: {
                    title: text
                },
                activeClass: ACTIVE_DROPDOWN_ITEM_CLASSNAME,
                disabledClass: DISABLED_CLASSNAME
            };
            var menuItemView = new MenuItemView(menuItem, _this.editorView, spec);
            var _a = menuItemView.render(), update = _a.update, dom = _a.dom;
            // remove open class once clicked on dropdown value
            dom.addEventListener('click', function (e) {
                e.preventDefault();
                closeDropdown();
            });
            // wrapper to execute when update is called
            var dropUpdate = function (state) {
                update(state);
                // update the dropdown content heading when a class is selected
                var activeEl = dropdownMenu.getElementsByClassName(ACTIVE_DROPDOWN_ITEM_CLASSNAME);
                if (activeEl.length) {
                    var el = activeEl[0];
                    dropdownText.textContent = el.textContent;
                    dropdown.classList.add(SELECTED_DROPDOWN_ITEM_CLASSNAME);
                }
                else {
                    // restore default value
                    dropdownText.textContent = labels[_this.dropdownGroup];
                    dropdown.classList.remove(SELECTED_DROPDOWN_ITEM_CLASSNAME);
                }
            };
            dropdownMenu.appendChild(dom);
            _this.updates.push(dropUpdate);
        });
        dropdown.appendChild(dropdownMenu);
        return dropdown;
    };
    DropDownView.prototype.render = function () {
        this.dom = this.getWrapperDom();
        return {
            dom: this.dom,
            updates: this.updates
        };
    };
    return DropDownView;
}());
var MenuItemView = /** @class */ (function () {
    function MenuItemView(menuItem, editorView, spec) {
        this.menuItem = menuItem;
        this.editorView = editorView;
        this.spec = spec;
    }
    MenuItemView.prototype.render = function () {
        var _this = this;
        var dom = this.dom = this.getDom();
        var schema = this.editorView.state.schema;
        var command = this.setupCommandListeners().command;
        var _a = this.spec, activeClass = _a.activeClass, disabledClass = _a.disabledClass;
        var update = function (state) {
            var menuItem = _this.menuItem;
            var isActive = false;
            var canExecute = command(_this.editorView.state, null);
            if (menuItem.type === 'mark') {
                var type = schema.marks[menuItem.key];
                isActive = isMarkActive(state, type);
            }
            if (menuItem.type === 'node') {
                var type = schema.nodes[menuItem.key];
                isActive = isNodeActive(state, type, menuItem.attrs);
            }
            dom.classList.toggle(activeClass, isActive);
            dom.classList.toggle(disabledClass, !canExecute);
        };
        return {
            dom: dom,
            update: update
        };
    };
    MenuItemView.prototype.getDom = function () {
        var div = document.createElement('div');
        if (this.spec.classNames) {
            this.spec.classNames.forEach(function (className) {
                div.classList.add(className);
            });
        }
        if (this.spec.attrs) {
            Object.entries(this.spec.attrs).forEach(function (obj) {
                div.setAttribute(obj[0], obj[1]);
            });
        }
        if (this.spec.innerHTML) {
            div.innerHTML = this.spec.innerHTML;
        }
        if (this.spec.textContent) {
            div.innerHTML = this.spec.textContent;
        }
        return div;
    };
    MenuItemView.prototype.setupCommandListeners = function () {
        var _this = this;
        var schema = this.editorView.state.schema;
        var command;
        if (this.menuItem.type === 'mark') {
            command = toggleMark(schema.marks[this.menuItem.key]);
        }
        if (this.menuItem.type === 'node') {
            var type = schema.nodes[this.menuItem.key];
            if (isListItem(type, schema)) {
                command = toggleList(type, schema.nodes.list_item);
            }
            if (type === schema.nodes.heading) {
                command = toggleBlockType(type, schema.nodes.paragraph, { level: this.menuItem.attrs.level });
            }
            if (type === schema.nodes.blockquote) {
                command = toggleWrap(type);
            }
        }
        this.dom.addEventListener('mousedown', function (e) {
            e.preventDefault();
            // don't execute if not left click
            if (e.buttons !== 1) {
                return;
            }
            // execute command
            command(_this.editorView.state, _this.editorView.dispatch);
        });
        return { command: command };
    };
    return MenuItemView;
}());
var getSeperatorDom = function () {
    var div = document.createElement('div');
    div.className = SEPERATOR_CLASSNAME;
    return div;
};
var ɵ0 = getSeperatorDom;
export var renderMenu = function (options, editorView, menuDom) {
    var updates = [];
    var toolbar = options.toolbar;
    toolbar.forEach(function (group, toolbarIndex) {
        var isLastMenuGroup = toolbar.length - 1 === toolbarIndex;
        group.forEach(function (toolbarItem, menuIndex) {
            var isLastMenuItem = group.length - 1 === menuIndex;
            // render dropdown
            if (typeof toolbarItem === 'object') {
                Object.keys(toolbarItem).forEach(function (dropdownGroup) {
                    if (DROPDOWN_ITEMS.has(dropdownGroup)) {
                        var dropdown = toolbarItem[dropdownGroup];
                        var dropdownView = new DropDownView(dropdownGroup, dropdown, editorView, options);
                        var rendered = dropdownView.render();
                        updates.push(rendered.updates);
                        menuDom.appendChild(rendered.dom);
                    }
                    else {
                        console.warn('Unkown dropdown group:', dropdownGroup);
                    }
                });
            }
            // render Icons
            if (typeof toolbarItem === 'string') {
                var menuItem = menuItemsMeta[toolbarItem];
                var labels = options.labels;
                if (menuItem) {
                    var spec = {
                        classNames: [
                            MENU_ITEM_CLASSNAME,
                            MENU_ITEM_CLASSNAME + "--Icon",
                        ],
                        innerHTML: getIconSvg(menuItem.icon),
                        attrs: {
                            title: labels[menuItem.i18nKey]
                        },
                        activeClass: ACTIVE_MENU_ITEM_CLASSNAME,
                        disabledClass: DISABLED_CLASSNAME
                    };
                    var menuItemView = new MenuItemView(menuItem, editorView, spec);
                    var _a = menuItemView.render(), update = _a.update, dom = _a.dom;
                    menuDom.appendChild(dom);
                    updates.push(update);
                }
            }
            if (typeof toolbarItem === 'function') {
                var _b = toolbarItem(editorView), dom = _b.dom, update = _b.update;
                menuDom.appendChild(dom);
                updates.push(update);
            }
            if (isLastMenuItem && !isLastMenuGroup) {
                var seperatorDom = getSeperatorDom();
                menuDom.appendChild(seperatorDom);
            }
        });
    });
    var combinedUpdates = flatDeep(updates, Infinity);
    return {
        update: function (state) {
            combinedUpdates.forEach(function (update) {
                update(state);
            });
        }
    };
};
export { ɵ0 };
